### PR TITLE
Weight ops for last summary

### PR DESCRIPTION
**Pull request by kian-thompson**
# [AB#3089](https://dev.azure.com/fluidframework/internal/_workitems/edit/3089)

[How contribute to this repo](https://github.com/microsoft/FluidFramework/blob/main/CONTRIBUTING.md).

[Guidelines for Pull Requests](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

## Description

There was a change in regular summary heuristics that weight runtime and non-runtime ops differently.

`shouldRunLastSummary` should also do something similar so that it doesn't end up summarizing too often which might be the case today. There were some changes relatively recently where `minOpsForLastSummaryAttempt` was changed to 10 and system ops might be contributing to this.

### CODE REVIEW

**Review by NicholasCouri — APPROVED** [review]

File: packages/runtime/container-runtime/src/summary/summarizerHeuristics.ts [hunk 1]
```diff
@@ -157,16 +157,21 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
 	}
 
 	public shouldRunLastSummary(): boolean {
-		const opsSinceLastAck = this.opsSinceLastAck;
+		const weightedOpsSinceLastAck = getWeightedNumberOfOps(
+			this.heuristicData.numRuntimeOps,
+			this.heuristicData.numNonRuntimeOps,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerHeuristics.ts:162]: Are nonRuntimeOps user ops or is the other way around?

File: packages/runtime/container-runtime/src/test/summary/summarizerHeuristics.spec.ts [hunk 2]
```diff
@@ -231,20 +231,18 @@ describe("Runtime", () => {
 			});
 
 			it("Should summarize on close if enough outstanding ops", () => {
-				const lastSummary = 1000;
 				const minOpsForLastSummaryAttempt = 10;
-				initialize({ refSequenceNumber: lastSummary, minOpsForLastSummaryAttempt });
+				initialize({ minOpsForLastSummaryAttempt, runtimeOpWeight: 1.0 });
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/summary/summarizerHeuristics.spec.ts:235]: You should add tests where both types of ops exist with different weights. Then validate that summarize on close runs accordingly.
For instance, 10 non-user ops should not trigger summary which is what would happen without this change.

**Review by agarwal-navin — APPROVED** [review]
Looks good. Please add more tests with both types of ops that validate that the behavior has infact changed with this change.

File: packages/runtime/container-runtime/src/summary/summarizerHeuristics.ts (hunk 1 above)
Comment (kian-thompson) [reply packages/runtime/container-runtime/src/summary/summarizerHeuristics.ts:162]: Runtime ops are the new "user" ops. This was changed in https://github.com/microsoft/FluidFramework/pull/10463

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/summarizerHeuristics.ts:162]: Unfortunately, that PR is not very descriptive and its hard to tell what happened, but I get the gist :-).

File: packages/runtime/container-runtime/src/test/summary/summarizerHeuristics.spec.ts [hunk 3]
```diff
@@ -231,21 +231,37 @@ describe("Runtime", () => {
...
+
+				data.numRuntimeOps = minOpsForLastSummaryAttempt - 1;
+				assert.strictEqual(runner.shouldRunLastSummary(), false, "should not run on close");
+			});
+
+			it("Should summarize on close weights ops properly", () => {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/summary/summarizerHeuristics.spec.ts:249]: Awesome!

**Pull request merged**